src/pgo/trans/intermediate/ModularPlusCalLabelingRulesVisitor.java

```diff
@@ -0,0 +1,398 @@
+package pgo.trans.intermediate;
+
+import pgo.Unreachable;
+import pgo.errors.IssueContext;
+import pgo.model.mpcal.ModularPlusCalYield;
+import pgo.model.pcal.*;
+import pgo.model.tla.TLAExpression;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.IntStream;
+
+/**
+ * Validates the labeling rules of a Modular PlusCal specification. The labeling rules
+ * used are exactly the same as those used by standard PlusCal and are described in
+ * its manual.
+ */
+public class ModularPlusCalLabelingRulesVisitor extends PlusCalStatementVisitor<Void, RuntimeException> {
+
+    /**
+     * Validates whether the next statement following a certain subtree in the AST requires to a label.
+     * This visitor assumes the root node of the AST is an If or Either expression (checks need to be
+     * performed by caller).
+     *
+     * The statement succeeding an 'if' or 'either' PlusCal statement needs to be labeled if there is:
+     *   - a label
+     *   - 'return'
+     *   - 'goto'
+     *   - 'call'
+     *
+     * anywhere within the 'if' or 'either' statement.
+     */
+    private class IfEitherNextStatementRequiresLabel extends PlusCalStatementVisitor<Boolean, RuntimeException> {
```

Comment:
What would be the advantage of that? This seems like a good use case for inner classes, since `IfEitherNextStatementRequiresLabel` implements a very specific type of check that will ever only be used by the outer class.